lib/propel/builder/QubitObjectBuilder.php

```diff
@@ -1924,6 +1929,46 @@ public function hasChildren()
     return (\$this->{$this->getColumnVarName($this->nestedSetRightColumn)} - \$this->{$this->getColumnVarName($this->nestedSetLeftColumn)}) > 1;
   }
 
+script;
+  }
+
+  protected function addGetAncestorsAndSelfForAcl(&$script)
+  {
+    $selfFkTableName = $this->selfFk->getTableName();
+    $selfFkColName = $this->selfFk->getLocalColumns()[0];
+    $selfFkRefColName = $this->selfFk->getForeignColumns()[0];
+    $selfFkRefColCons = $this->getColumnConstant($this->getTable()->getColumn($selfFkRefColName));
+
+    $script .= <<<script
+
+  public function getAncestorsAndSelfForAcl()
+  {
+    if (!isset(\$this->values['ancestorsAndSelfForAcl']))
+    {
+      \$cte = "(
+      	WITH RECURSIVE aas AS
+      	(
+      	  SELECT tb1.{$selfFkRefColName}, tb1.{$selfFkColName}, 1 as lev
+          FROM {$selfFkTableName} tb1
+          WHERE tb1.{$selfFkRefColName}=\$this->id
+      	  UNION ALL
+      	  SELECT tb2.{$selfFkRefColName}, tb2.{$selfFkColName}, aas.lev + 1
+          FROM {$selfFkTableName} tb2
+          JOIN aas ON aas.{$selfFkColName}=tb2.{$selfFkRefColName}
+      	)
+      	SELECT {$selfFkRefColName}, lev FROM aas
+      )";
+
+      \$criteria = new Criteria;
+      \$criteria->addJoin({$selfFkRefColCons}, 'cte.id', "RIGHT JOIN \$cte");
```

Comment:
Huh, this is cool - I didn't realize you could use raw SQL in a Propel Criteria join like this. 😮

Reply:
Me neither :smile:! It wasn't easy to figure out and I feel like I'm hacking it, but it works :stuck_out_tongue:.

Reply:
This is going to really help the digital object browse page!

Reply:
Thanks @sbreker!

We were already using CTE but it was in a subquery so this should be a bit better. Additionally, it removes the extra processing from `andSelf()` and `orderBy()`:

https://github.com/artefactual/atom/blob/qa/2.6.x/lib/QubitQuery.class.php#L137-L170

:rocket: